File: server/notification/github/src/main/java/projektor/notification/github/JwtProvider.java
package projektor.notification.github;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.PEMParser;
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;

import java.io.StringReader;
import java.security.KeyPair;
import java.security.PrivateKey;
import java.security.Security;
import java.security.interfaces.RSAPrivateKey;
import java.util.Date;

public class JwtProvider {
    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    String createJWT(String githubAppId, String pemKey, long ttlMillis) throws Exception {
        RSAPrivateKey privateKey = getPemPrivateKey(pemKey);
        Algorithm algorithm = Algorithm.RSA256(null, privateKey);
        long nowMillis = System.currentTimeMillis();
        Date now = new Date(nowMillis);

        long expMillis = nowMillis + ttlMillis;
        Date exp = new Date(expMillis);

        String token = JWT.create()
                .withIssuer(githubAppId)
                .withIssuedAt(now)
                .withExpiresAt(exp)
                .sign(algorithm);

        return token;
    }

    public static RSAPrivateKey getPemPrivateKey(String mKey) throws Exception {
        PEMParser pemParser = new PEMParser(new StringReader(mKey));
        JcaPEMKeyConverter converter = new JcaPEMKeyConverter().setProvider("BC");
        Object object = pemParser.readObject();
        KeyPair kp = converter.getKeyPair((PEMKeyPair) object);
        PrivateKey privateKey = kp.getPrivate();
        return (RSAPrivateKey)privateKey;
    }
}

File: server/notification/github/src/main/kotlin/projektor/notification/github/auth/JwtProvider.kt
package projektor.notification.github.auth

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.bouncycastle.openssl.PEMKeyPair
import org.bouncycastle.openssl.PEMParser
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter
import java.io.StringReader
import java.security.Security
import java.security.interfaces.RSAPrivateKey
import java.util.Date

open class JwtProvider(private val jwtTokenConfig: JwtTokenConfig) {
    init {
        Security.addProvider(BouncyCastleProvider())
    }

    open fun createJWT(): String {
        val privateKey = getPemPrivateKey(jwtTokenConfig.pemContents)
        val algorithm = Algorithm.RSA256(null, privateKey)
        val nowMillis = System.currentTimeMillis()
        val now = Date(nowMillis)

        val expMillis: Long = nowMillis + jwtTokenConfig.ttlMillis
        val exp = Date(expMillis)

        return JWT.create()
            .withIssuer(jwtTokenConfig.gitHubAppId)
            .withIssuedAt(now)
            .withExpiresAt(exp)
            .sign(algorithm)
    }

    companion object {
        private fun getPemPrivateKey(mKey: String): RSAPrivateKey {
            val pemParser = PEMParser(StringReader(mKey))
            val converter = JcaPEMKeyConverter().setProvider("BC")
            val pemKeyPair = pemParser.readObject() as PEMKeyPair
            val kp = converter.getKeyPair(pemKeyPair)
            val privateKey = kp.private
            return privateKey as RSAPrivateKey
        }
    }
}
